src/main/java/org/fisco/bcos/sdk/service/GroupManagerServiceImpl.java

```diff
@@ -0,0 +1,380 @@
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fisco.bcos.sdk.service;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.commons.lang3.StringUtils;
+import org.fisco.bcos.sdk.channel.Channel;
+import org.fisco.bcos.sdk.channel.PeerSelectRule;
+import org.fisco.bcos.sdk.channel.ResponseCallback;
+import org.fisco.bcos.sdk.channel.model.Options;
+import org.fisco.bcos.sdk.client.Client;
+import org.fisco.bcos.sdk.client.exceptions.ClientException;
+import org.fisco.bcos.sdk.client.protocol.response.GroupList;
+import org.fisco.bcos.sdk.model.Message;
+import org.fisco.bcos.sdk.model.Response;
+import org.fisco.bcos.sdk.network.ConnectionInfo;
+import org.fisco.bcos.sdk.service.model.BlockNumberMessageDecoder;
+import org.fisco.bcos.sdk.service.model.BlockNumberNotification;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GroupManagerServiceImpl implements GroupManagerService {
+    private static Logger logger = LoggerFactory.getLogger(GroupManagerServiceImpl.class);
+    private final Channel channel;
+    private final GroupServiceFactory groupServiceFactory;
+    private ConcurrentHashMap<Integer, GroupService> groupIdToService = new ConcurrentHashMap<>();
+    private ConcurrentHashMap<String, List<String>> nodeToGroupIDList = new ConcurrentHashMap<>();
+    private final BlockNumberMessageDecoder blockNumberMessageDecoder;
+    private Client groupInfoGetter;
+    // TODO: get the fetchGroupListIntervalMs from the configuration
+    private long fetchGroupListIntervalMs = 60000;
+    private AtomicBoolean running;
+    private Thread worker;
+
+    public GroupManagerServiceImpl(Channel channel) {
+        this.channel = channel;
+        this.groupServiceFactory = new GroupServiceFactory();
+        this.groupInfoGetter = Client.build(channel);
+        fetchGroupList();
+        this.blockNumberMessageDecoder = new BlockNumberMessageDecoder(channel.getVersion());
+        this.start();
+    }
+
+    @Override
+    public Channel getChannel() {
+        return this.channel;
+    }
+
+    protected void finalize() {
+        stop();
+    }
+
+    /** Stop group list fetching thread */
+    protected void stop() {
+        if (running.get()) {
+            logger.warn("GroupManagerService has already been stopped, ");
+            return;
+        }
+        running.set(false);
+        this.worker.interrupt();
+    }
+
+    /** start the thread to obtain group list information periodically */
+    protected void start() {
+        if (this.worker != null) {
+            logger.warn("GroupManagerService has already been started");
+        }
```

Comment:
Why not use running to determine whether this service is running?

Reply:
done.